lib/cache/OperationWriter.js

```diff
@@ -0,0 +1,255 @@
+   */
+  async add({operation, meta, forceFlush = false} = {}) {
+    const {queue, running, highWaterMark, MAX_OPERATION_QUEUE_LENGTH} = this;
+
+    if(this.databaseOperationQueueLength >= MAX_OPERATION_QUEUE_LENGTH) {
+      // we intentionally do not await the promise, the length can be updated
+      // after the error is returned
+      this._cacheDatabaseQueueLength().catch(e => {
+        logger.debug(DATABASE_QUEUE_LENGTH_ERROR, {error: e});
+      });
+      throw new BedrockError(
+        'The node is not accepting operations. Try again later.',
+        'OperationError', {
+          httpStatusCode: 503,
+          public: true,
+        });
+    }
+
+    if(this._isDuplicateOperation(meta)) {
+      throw new BedrockError(
+        'The operation already exists.',
+        'DuplicateError', {
+          duplicateLocation: 'cache',
+          httpStatusCode: 409,
+          ledgerNodeId: this.ledgerNodeId,
+          operationRecord: {operation, meta},
+          public: true
+        });
+    }
+
+    // queue the operation
+    queue.push({operation, meta});
+
+    // do not flush if not writer is not running
+    if(!running) {
+      return;
+    }
+
+    // force flush if required
+    if(forceFlush) {
+      await this.flush();
+    }
+
+    // if the queue length matches the highWaterMark, schedule an immediate
+    // flush; only schedule this when they match do prevent repeatedly
+    // scheduling flushes or preventing flushes under heavy load
+    if(queue.length === highWaterMark) {
+      this._scheduleFlush({timeout: 0});
+      return;
+    }
+
+    // schedule a flush if one has not already been scheduled
+    const {timeout, timeoutId} = this;
+    if(timeoutId === null) {
+      this._scheduleFlush({timeout});
+    }
+  }
+
+  /**
+   * Forcibly flushes the operation queue, bulk inserting any pending
+   * operations.
+   *
+   * @returns {Promise} A promise that resolves once the operation completes.
+   */
+  async flush() {
+    // clear any scheduled flush
+    if(this.timeoutId !== null) {
+      clearTimeout(this.timeoutId);
+      this.timeoutId = null;
+    }
+
+    // save current queue and replace it with a new one
+    const queue = this.queue;
+    this.queue = [];
+
+    if(queue.length === 0) {
+      return;
+    }
+
+    const now = Date.now();
+    const records = queue.map(({meta, operation}) => ({
+      ledgerNodeId: this.ledgerNodeId,
+      meta: {
+        ...meta,
+        created: now,
+        updated: now
+      },
+      operation
+    }));
+
+    try {
+      const writeOptions = {...database.writeOptions, ordered: false};
+      await this.collection.insertMany(records, writeOptions);
+      await this._cacheDatabaseQueueLength().catch(e => {
+        logger.debug(DATABASE_QUEUE_LENGTH_ERROR, {error: e});
+      });
+    } catch(e) {
+      if(!database.isDuplicateError(e)) {
+        throw e;
+      }
+    }
+  }
+
+  /**
+   * Caches the length of the database in queue in memory.
```

Comment:
```suggestion
   * Updates the length of the database in queue in memory if it hasn't been
   * updated too recently. Updates to the database queue length are debounced
   * for better performance. Getting a consistent value for the database queue
   * length is also not part of the design, as multiple isolated processes may be
   * writing to the database queue concurrently and another isolated process
   * may be draining it concurrently. Therefore, design here presumes there
   * will be not strong consistency when tracking the database length; rather,
   * it allows for some operations to be refused to be queued when under load,
   * even if it may have been technically possible for those operations to be
   * queued with a strongly consistent design. This is considered to be an
   * acceptable condition given the performance benefits from not requiring
   * strong consistency and with the observation that false positives will be
   * in the margins on a taxed system.
```